src/main/java/com/amazon/opendistroforelasticsearch/sql/optimizer/subquery/InSubqueryRewriter.java

```diff
@@ -0,0 +1,146 @@
+        this.inSubQuery = (MySqlSelectQueryBlock) inSubQueryExpr.getSubQuery().getQuery();
+        joinLeftExpr = inSubQueryExpr.getExpr();
+        joinRightExpr = rightExpr(inSubQuery);
+    }
+
+    /**
+     * Rewrite the subquery with JOIN clause. This API is NOT idempotent.
+     *
+     * @param query Subquery will be written.
+     */
+    @Override
+    public void rewrite(MySqlSelectQueryBlock query) {
+        query.setFrom(buildFrom(query));
+        query.setWhere(buildWhere(query));
+    }
+
+    /**
+     * Build From clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FROM
+     *                         |     |
+     *                         b     B
+     *
+     * The FROM logic should be
+     *     JOIN
+     *     / | \
+     *    A  B A.a == B.b
+     */
+    private SQLJoinTableSource buildFrom(MySqlSelectQueryBlock rootQuery) {
+        SQLJoinTableSource sqlJoin = new SQLJoinTableSource();
+        sqlJoin.setLeft(rootQuery.getFrom());
+        sqlJoin.setRight(inSubQuery.getFrom());
+        sqlJoin.setCondition(new SQLBinaryOpExpr(joinLeftExpr, SQLBinaryOperator.Equality, joinRightExpr));
+        sqlJoin.setJoinType(JoinType.JOIN);
+
+        return sqlJoin;
+    }
+
+    /**
+     * Build Where clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FROM
+     *                         |     |
+     *                         b     B
+     *
+     *
+     *     WHERE
+     *     /  |   \
+     *   c>10 AND B.b is NOT NULL
+     */
+    private SQLExpr buildWhere(MySqlSelectQueryBlock rootQuery) {
+        return rewriteSubquery(rootQuery.getWhere());
+    }
+
+    private SQLExpr rewriteSubquery(SQLExpr expr) {
+        if (expr instanceof SQLInSubQueryExpr) {
+            return createSubqueryReplacementCondition();
+        }
+        else if (expr instanceof SQLBinaryOpExpr) {
+            ((SQLBinaryOpExpr) expr).setLeft(rewriteSubquery(((SQLBinaryOpExpr) expr).getLeft()));
+            ((SQLBinaryOpExpr) expr).setRight(rewriteSubquery(((SQLBinaryOpExpr) expr).getRight()));
+            return expr;
+        } else {
+            return expr;
+        }
+    }
+
+    private SQLExpr rightExpr(MySqlSelectQueryBlock rightQuery) {
+        if (rightQuery.getSelectList().size() > 1) {
+            throw new IllegalStateException("Unsupported subquery with multiple select " + rightQuery.getSelectList());
```

Comment:
Do you consider this case being a bug in our code? If it's possible to get to this from invalid user input, then you should consider changing the exception type to one of the exceptions that are considered client errors (see in RestSqlAction.java).